1) Prepare a classification model using SVM for salary data 

Data Description:

age -- age of a person

workclass	-- A work class is a grouping of work 

education	-- Education of an individuals	

maritalstatus -- Marital status of an individulas	

occupation	 -- occupation of an individuals

relationship -- 	
race --  Race of an Individual

sex --  Gender of an Individual

capitalgain --  profit received from the sale of an investment	

capitalloss	-- A decrease in the value of a capital asset

hoursperweek -- number of hours work per week	

native -- Native of an individual

Salary -- salary of an individual

In [2]:
# SVM Classification
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler

from sklearn import svm
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report


from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn import preprocessing

In [3]:
data= pd.read_csv("SalaryData_Train(1).csv")

In [4]:
data.head()

age          workclass   education  educationno        maritalstatus  \
0   39          State-gov   Bachelors           13        Never-married   
1   50   Self-emp-not-inc   Bachelors           13   Married-civ-spouse   
2   38            Private     HS-grad            9             Divorced   
3   53            Private        11th            7   Married-civ-spouse   
4   28            Private   Bachelors           13   Married-civ-spouse   

           occupation    relationship    race      sex  capitalgain  \
0        Adm-clerical   Not-in-family   White     Male         2174   
1     Exec-managerial         Husband   White     Male            0   
2   Handlers-cleaners   Not-in-family   White     Male            0   
3   Handlers-cleaners         Husband   Black     Male            0   
4      Prof-specialty            Wife   Black   Female            0   

   capitalloss  hoursperweek          native  Salary  
0            0            40   United-States   <=50K  
1            0            13   United-States   <=50K  
2            0            40   United-States   <=50K  
3            0            40   United-States   <=50K  
4            0            40            Cuba   <=50K

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30161 entries, 0 to 30160
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   age            30161 non-null  int64 
 1   workclass      30161 non-null  object
 2   education      30161 non-null  object
 3   educationno    30161 non-null  int64 
 4   maritalstatus  30161 non-null  object
 5   occupation     30161 non-null  object
 6   relationship   30161 non-null  object
 7   race           30161 non-null  object
 8   sex            30161 non-null  object
 9   capitalgain    30161 non-null  int64 
 10  capitalloss    30161 non-null  int64 
 11  hoursperweek   30161 non-null  int64 
 12  native         30161 non-null  object
 13  Salary         30161 non-null  object
dtypes: int64(5), object(9)
memory usage: 3.2+ MB


In [6]:
data.columns

Index(['age', 'workclass', 'education', 'educationno', 'maritalstatus',
       'occupation', 'relationship', 'race', 'sex', 'capitalgain',
       'capitalloss', 'hoursperweek', 'native', 'Salary'],
      dtype='object')

In [7]:
a= ['workclass', 'education',  'maritalstatus', 'occupation', 'relationship', 'race', 'sex', 'native', 'Salary']


In [8]:
label_encoder = preprocessing.LabelEncoder()
for i in a:
    data[i]= label_encoder.fit_transform(data[i]) 

In [9]:
data.head()

age  workclass  education  educationno  maritalstatus  occupation  \
0   39          5          9           13              4           0   
1   50          4          9           13              2           3   
2   38          2         11            9              0           5   
3   53          2          1            7              2           5   
4   28          2          9           13              2           9   

   relationship  race  sex  capitalgain  capitalloss  hoursperweek  native  \
0             1     4    1         2174            0            40      37   
1             0     4    1            0            0            13      37   
2             1     4    1            0            0            40      37   
3             0     2    1            0            0            40      37   
4             5     2    0            0            0            40       4   

   Salary  
0       0  
1       0  
2       0  
3       0  
4       0

In [10]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30161 entries, 0 to 30160
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype
---  ------         --------------  -----
 0   age            30161 non-null  int64
 1   workclass      30161 non-null  int64
 2   education      30161 non-null  int64
 3   educationno    30161 non-null  int64
 4   maritalstatus  30161 non-null  int64
 5   occupation     30161 non-null  int64
 6   relationship   30161 non-null  int64
 7   race           30161 non-null  int64
 8   sex            30161 non-null  int64
 9   capitalgain    30161 non-null  int64
 10  capitalloss    30161 non-null  int64
 11  hoursperweek   30161 non-null  int64
 12  native         30161 non-null  int64
 13  Salary         30161 non-null  int64
dtypes: int64(14)
memory usage: 3.2 MB


In [11]:
data.Salary.value_counts()

0    22653
1     7508
Name: Salary, dtype: int64

In [12]:
data.describe()

age    workclass    education   educationno  maritalstatus  \
count  30161.000000  30161.00000  30161.00000  30161.000000   30161.000000   
mean      38.438115      2.19933     10.33361     10.121316       2.580087   
std       13.134830      0.95394      3.81226      2.550037       1.498018   
min       17.000000      0.00000      0.00000      1.000000       0.000000   
25%       28.000000      2.00000      9.00000      9.000000       2.000000   
50%       37.000000      2.00000     11.00000     10.000000       2.000000   
75%       47.000000      2.00000     12.00000     13.000000       4.000000   
max       90.000000      6.00000     15.00000     16.000000       6.000000   

         occupation  relationship          race           sex   capitalgain  \
count  30161.000000  30161.000000  30161.000000  30161.000000  30161.000000   
mean       5.959849      1.418322      3.678592      0.675707   1092.044064   
std        4.029633      1.601361      0.834721      0.468118   7406.466611   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        2.000000      0.000000      4.000000      0.000000      0.000000   
50%        6.000000      1.000000      4.000000      1.000000      0.000000   
75%        9.000000      3.000000      4.000000      1.000000      0.000000   
max       13.000000      5.000000      4.000000      1.000000  99999.000000   

        capitalloss  hoursperweek        native        Salary  
count  30161.000000  30161.000000  30161.000000  30161.000000  
mean      88.302311     40.931269     35.413481      0.248931  
std      404.121321     11.980182      5.956430      0.432401  
min        0.000000      1.000000      0.000000      0.000000  
25%        0.000000     40.000000     37.000000      0.000000  
50%        0.000000     40.000000     37.000000      0.000000  
75%        0.000000     45.000000     37.000000      0.000000  
max     4356.000000     99.000000     39.000000      1.000000

In [13]:
n= min(data.shape)

In [14]:
n

14

In [15]:
array = data.values
X = array[:,0:n]
Y = array[:,-1]

In [16]:
X

array([[39,  5,  9, ..., 40, 37,  0],
       [50,  4,  9, ..., 13, 37,  0],
       [38,  2, 11, ..., 40, 37,  0],
       ...,
       [58,  2, 11, ..., 40, 37,  0],
       [22,  2, 11, ..., 20, 37,  0],
       [52,  3, 11, ..., 40, 37,  1]])

In [17]:
Y

array([0, 0, 0, ..., 0, 0, 1])

In [18]:
X_train =X

In [19]:
X_train

array([[39,  5,  9, ..., 40, 37,  0],
       [50,  4,  9, ..., 13, 37,  0],
       [38,  2, 11, ..., 40, 37,  0],
       ...,
       [58,  2, 11, ..., 40, 37,  0],
       [22,  2, 11, ..., 20, 37,  0],
       [52,  3, 11, ..., 40, 37,  1]])

In [20]:
Y_train = Y

### Grid Search CV

In [ ]:
# clf = SVC()
# param_grid = [{'kernel':['rbf'],'gamma':[10,0.5],'C':[10,0.1] }]
# gsv = GridSearchCV(clf,param_grid,cv=10)
# gsv.fit(X_train,Y_train)

In [33]:
# gsv.best_params_ , gsv.best_score_ 

In [22]:
clf = SVC(C= 15, gamma = 0.5)
clf.fit(X_train , Y_train)

SVC(C=15, gamma=0.5)

In [23]:
data1 = pd.read_csv("SalaryData_Test(1).csv")

In [24]:
data1.head()

age   workclass      education  educationno        maritalstatus  \
0   25     Private           11th            7        Never-married   
1   38     Private        HS-grad            9   Married-civ-spouse   
2   28   Local-gov     Assoc-acdm           12   Married-civ-spouse   
3   44     Private   Some-college           10   Married-civ-spouse   
4   34     Private           10th            6        Never-married   

           occupation    relationship    race    sex  capitalgain  \
0   Machine-op-inspct       Own-child   Black   Male            0   
1     Farming-fishing         Husband   White   Male            0   
2     Protective-serv         Husband   White   Male            0   
3   Machine-op-inspct         Husband   Black   Male         7688   
4       Other-service   Not-in-family   White   Male            0   

   capitalloss  hoursperweek          native  Salary  
0            0            40   United-States   <=50K  
1            0            50   United-States   <=50K  
2            0            40   United-States    >50K  
3            0            40   United-States    >50K  
4            0            30   United-States   <=50K

In [25]:
label_encoder = preprocessing.LabelEncoder()
for i in a:
    data1[i]= label_encoder.fit_transform(data1[i]) 

In [26]:
data1.head()

age  workclass  education  educationno  maritalstatus  occupation  \
0   25          2          1            7              4           6   
1   38          2         11            9              2           4   
2   28          1          7           12              2          10   
3   44          2         15           10              2           6   
4   34          2          0            6              4           7   

   relationship  race  sex  capitalgain  capitalloss  hoursperweek  native  \
0             3     2    1            0            0            40      37   
1             0     4    1            0            0            50      37   
2             0     4    1            0            0            40      37   
3             0     2    1         7688            0            40      37   
4             1     4    1            0            0            30      37   

   Salary  
0       0  
1       0  
2       1  
3       1  
4       0

In [31]:
array1 = data1.values
X_test= array1[:,0:n]
y_test = array1[:,-1]

In [32]:
y_pred = clf.predict(X_test)
acc = accuracy_score(y_test, y_pred) * 100
print("Accuracy =", acc)
confusion_matrix(y_test, y_pred)

Accuracy = 83.22709163346613


array([[11231,   129],
       [ 2397,  1303]])

In [34]:
y_pred

array([0, 0, 1, ..., 0, 0, 0])

### In the above analysis, we used label 0 for the salary <=50K, and label 1 for salary >50k. We find the SVM based model with the 83.2% accuracy here with C = 15, gamma = 0.5 in RBF kernal fuction. Here, C is used for soft margin ( low c - allow soft margin, high c - punish model badly if make mistake), higher gamma signigy the more complexity in model by cosidering the more influcence of the dimensions on the deciosin bounday.